File: Segment Tree/1526. Minimum Number of Increments on Subarrays to Form a Target Array/solution.py
# Segment Tree + Recursion
class Seg():
    def __init__(self, nums, op=min):
        n = len(nums)
        self.op = op
        self.n = n
        self.tree = [[0, 0] for _ in range(2*n)]


        for i in range(n, 2*n):
            self.tree[i][0] = nums[i-n]
            self.tree[i][1] = i-n

        # 1-based
        for i in range(n-1, 0, -1):
            self.tree[i][0] = op(self.tree[i<<1][0], self.tree[i<<1|1][0])
            self.tree[i][1] = self.tree[i<<1][1] if self.tree[i<<1][0] == self.tree[i][0] else self.tree[i<<1|1][1]

    def query(self, l, r):
        tree, op, n = self.tree, self.op, self.n
        l, r = l+n, r+n

        res = inf
        pos = -1
        while l <= r:
            if l == r:
                res = op(res, tree[l][0])
                if tree[l][0] == res:
                    pos = tree[l][1]
                break

            if l&1 == 1:
                res = op(res, tree[l][0])
                if tree[l][0] == res:
                    pos = tree[l][1]
                l += 1

            if r&1 == 0:
                res = op(res, tree[r][0])
                if tree[r][0] == res:
                    pos = tree[r][1]
                r -= 1
            l, r = l>>1, r>>1
        return res, pos

class Solution:
    def minNumberOperations(self, target: List[int]) -> int:
        n = len(target)
        seg = Seg(target)

        def dfs(l, r, base):
            if l > r: return 0
            if l == r: return target[l]-base

            rangeMin, idx = seg.query(l, r)

            inc = rangeMin-base
            inc += dfs(l, idx-1, rangeMin)
            inc += dfs(idx+1, r, rangeMin)
            return inc
            
        return dfs(0, n-1, 0)


# Greedy
class Solution:
    def minNumberOperations(self, target: List[int]) -> int:
        n = len(target)
        res, currHeight = 0, 0
        for i in range(n):
            if target[i] > currHeight:
                res += target[i]-currHeight
            currHeight = target[i]
        return res

File: Segment Tree/1526. Minimum Number of Increments on Subarrays to Form a Target Array/solution.cpp
// Greedy: check adjacent difference
class Solution {
public:
    int minNumberOperations(vector<int>& target) {
        int n = target.size();
        int res = 0, currentHeight = 0;
        for (int i=0; i<n; i++) {
            if (target[i] > currentHeight) {
                res += target[i] - currentHeight;
            }

            currentHeight = target[i];
        }
        return res;
    }
};

// segment tree
class Solution {
private:
    class SegTreeNode
    {
        public:
        SegTreeNode* left = NULL;
        SegTreeNode* right = NULL;
        int start, end;
        int info, pos;  // the maximum value of the range
        bool tag;

        
        SegTreeNode(int a, int b, vector<int>& val)  // init for range [a,b] with the same-size array val
        {                 
            tag = 0;
            info = 0;
            start = a, end = b;
            if (a==b)
            {
                info = val[a];
                pos = a;
                return;
            }        
            int mid = (a+b)/2;
            if (left==NULL)
            {
                left = new SegTreeNode(a, mid, val);
                right = new SegTreeNode(mid+1, b, val);            
                if (left->info < right->info) {
                    info = left->info;
                    pos = left->pos;
                } else {
                    info = right->info;
                    pos = right->pos;
                }
            }        
        } 
        
        void pushDown()
        {
            if (tag==1 && left)
            {
                left->info = info;
                right->info = info;
                left->pos = pos;
                right->pos = pos;
                left->tag = 1;
                right->tag = 1;
                tag = 0;
            }        
        } 
        
        pair<int,int> queryRange(int a, int b)     // query the maximum value within range [a,b]
        {
            if (b < start || a > end )
            {
                return {INT_MAX/2,-1};  // check with your own logic
            }
            if (a <= start && end <=b)
            {
                return {info,pos};  // check with your own logic
            }          
            
            if (left)
            {
                pushDown();   
                auto L = left->queryRange(a, b);
                auto R = right->queryRange(a, b);

                if (L.first < R.first) {
                    return L;
                } else {
                    return R;
                }
            }
            
            return {info, pos};   // should not reach here
        }  

    };
public:
    int minNumberOperations(vector<int>& target) {
        int n = target.size();
        SegTreeNode* root = new SegTreeNode(0, n-1, target);
        
        function<int(int,int,int)> dfs = [&](int l, int r, int base) -> int {
            if (l > r) return 0;
            if (l == r) return target[l] - base;

            auto [val, pos] = root->queryRange(l, r);

            int sum = val - base;
            sum += dfs(l, pos-1, val);
            sum += dfs(pos+1, r, val);
            return sum;
        };

        return dfs(0, n-1, 0);
    }
};
